### PR TITLE
Functional Tests for Git-Flow Feature

### PR DESCRIPTION
**Pull request by FenrirUnbound**
Rewritten the functional tests for the Git Flow feature. This asserts the scenarios described in the `git-flow.feature` file. 

- [x] Scenario: New Pull Request
- [x] Updated Pull Request
- [x] Closed Pull Request
- [x] New Commit

### CODE REVIEW

File: features/support/github.js [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+ * Close a pull request for a given repository
+ * @method closePullRequest
+ * @param  {String}     token              Github token
+ * @param  {String}     repoOwner          Owner of the repository
+ * @param  {String}     repoName           Name of the repository
+ * @param  {Number}     prNumber           Number of the pull request
+ * @return {Promise}
+ */
+function closePullRequest(token, repoOwner, repoName, prNumber) {
+    return github.pullRequests.update({
```
Comment (tkyi) [inline features/support/github.js]: Do we also need 
```
github.authenticate({
         type: 'oauth',
         token
     });
```
here?

File: features/support/github.js [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+ * Create a branch on the given repository
+ * @method createBranch
+ * @param  {String}     token              Github token
+ * @param  {String}     branch             Name of the branch to create
+ * @param  {String}     [repoOwner]        Owner of the repository
+ * @param  {String}     [repoName]         Name of the repository
+ * @return {Promise}
+ */
+function createBranch(token, branch, repoOwner, repoName) {
+    const user = repoOwner || 'screwdriver-cd';
```
Comment (d2lam) [inline features/support/github.js]: `screwdriver-cd-test`

File: features/support/github.js [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+    const repo = repoName || 'garbage-repository-ignore-this';
```
Comment (d2lam) [inline features/support/github.js]: `functional-git`

File: features/support/github.js [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+        state: 'closed'
+    });
+}
+
+/**
+ * Create a branch on the given repository
+ * @method createBranch
+ * @param  {String}     token              Github token
+ * @param  {String}     branch             Name of the branch to create
+ * @param  {String}     [repoOwner]        Owner of the repository
```
Comment (d2lam) [inline features/support/github.js]: Should we just make these required? You make them required in `createFile`

File: features/support/github.js [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+ * @method createPullRequest
+ * @param  {String}   token             Github token
+ * @param  {String}   branch            The branch to create the file in
+ * @param  {String}   repoOwner         Owner of the repository
+ * @param  {String}   repoName          Name of the repository
+ * @return {Promise}
+ */
+function createPullRequest(token, branch, repoOwner, repoName) {
+    const repo = repoName;
+    const user = repoOwner;
```
Comment (d2lam) [inline features/support/github.js]: Don't think you need to create new variables here. They are only used once.

File: features/support/github.js (hunk 4 above)
Comment (FenrirUnbound) [inline features/support/github.js]: They get default values if you don't provide them. So optional.

File: features/support/github.js (hunk 5 above)
Comment (FenrirUnbound) [inline features/support/github.js]: Saying `repoOwner` is much more distinct than saying `user` in our API. The new variable names are mainly for having better usability with the `github` module. We cannot control its API, but we can control ours.

File: features/support/github.js (hunk 2 above)
Comment (FenrirUnbound) [inline features/support/github.js]: Amended

File: features/support/github.js (hunk 3 above)
Comment (FenrirUnbound) [inline features/support/github.js]: Amended

File: features/support/github.js (hunk 4 above)
Comment (d2lam) [inline features/support/github.js]: I mean, why do you make them optional here but required in `createFile`: https://github.com/screwdriver-cd/screwdriver/blob/575fd0205959a58b7aaeb040eda6ea54e2f72f3d/features/support/github.js#L116-L117

They are very similar functions so should keep them consistent.

File: features/support/github.js (hunk 5 above)
Comment (d2lam) [inline features/support/github.js]: What do you mean? Can't you just put:
```js
return github.pullRequests.create({
   user: repoOwner, 
   repo: repoName, 
   ...
});
```
Just like this one: https://github.com/screwdriver-cd/screwdriver/blob/575fd0205959a58b7aaeb040eda6ea54e2f72f3d/features/support/github.js#L184-L188

File: features/support/github.js (hunk 4 above)
Comment (tkyi) [inline features/support/github.js]: Addressed

File: features/step_definitions/git-flow.js [hunk 6]
```diff
@@ -104,90 +16,44 @@ function promiseToWait(timeToWait) {
...
-        this.testBranch = 'testBranch';
-        const params = {
-            user: this.repositoryOwner,
-            repo: this.repository,
-            ref: `heads/${this.testBranch}`
-        };
+        this.branch = 'darrenBranch';
+        this.repoOrg = 'screwdriver-cd-test';
+        this.repoName = 'functional-git';
+        this.pipelineId = '2e0138dfa7c4ff83720dc0cd510d2252a3398fc3';  // TODO: determine dynamically
```
Comment (minz1027) [inline features/step_definitions/git-flow.js]: Nice to know you can set up the Before hook for a given tag! But we already have `this.instance` in the before-hook under support. Do we really need it here?

File: features/step_definitions/git-flow.js [hunk 7]
```diff
@@ -248,39 +101,65 @@ module.exports = function server() {
...
-
-        return github.pullRequests.update({
-            user: this.repositoryOwner,
-            repo: this.repository,
-            number: this.pullRequestNumber,
-            state: 'closed'
-        });
-    });
+    this.When(/^the pull request is closed$/, {
+        timeout: 60 * 1000
```
Comment (minz1027) [inline features/step_definitions/git-flow.js]: Saw the timeout has been set in` this.Before` as 60000. Still need it here?

Comment (d2lam) [inline features/step_definitions/git-flow.js]: Yeah the timeout option in before doesn't get applied here. It needs to define at every scenario.

Comment (minz1027) [inline features/step_definitions/git-flow.js]: Okay, got it. Seems like the timeout here is only for async call inside the before hook

**Comment by d2lam** [comment]
Let's merge this after https://github.com/screwdriver-cd/screwdriver/pull/243

File: screwdriver.yaml [hunk 8]
```diff
@@ -68,5 +68,9 @@ jobs:
             K8S_DEPLOYMENT: sdapi
             SD_API: api.screwdriver.cd
         secrets:
+            # Access key for functional tests
+            - ACCESS_KEY
```
Comment (d2lam) [inline screwdriver.yaml]: I think this should be `SECRET_ACCESS_KEY`, but https://github.com/screwdriver-cd/screwdriver/pull/243 should fix it.

File: features/step_definitions/git-flow.js [hunk 9]
```diff
@@ -104,90 +16,44 @@ function promiseToWait(timeToWait) {
...
-        });
+        // Reset shared information
+        this.pullRequestNumber = null;
 
-        return github.gitdata.getReference(params)
-            // If branch exists, delete it
-            .then(() => github.gitdata.deleteReference(params), () => {});
+        return request({  // TODO : perform this in the before-hook for all func tests
+            method: 'GET',
+            url: `${this.instance}/v4/auth/token?access_key=${this.accessKey}`,
```
Comment (tkyi) [inline features/step_definitions/git-flow.js]: Addressed

File: features/step_definitions/git-flow.js (hunk 6 above)
Comment (tkyi) [inline features/step_definitions/git-flow.js]: fixed

File: features/step_definitions/git-flow.js [hunk 10]
```diff
@@ -104,90 +16,43 @@ function promiseToWait(timeToWait) {
...
 
-    this.Given(/^an existing pipeline$/, () => {
+    this.Given(/^an existing pipeline$/, () =>
         request({
-            uri: `${this.instance}/v4/pipelines`,
+            uri: `${this.instance}/${this.namespace}/pipelines`,
             method: 'POST',
             auth: {
-                username: this.username,
+                user: this.username,
```
Comment (minz1027) [inline features/step_definitions/git-flow.js]: Don't think we need `username` here. `jwt` alone here should be sufficient.

File: features/step_definitions/git-flow.js [hunk 11]
```diff
@@ -104,90 +16,44 @@ function promiseToWait(timeToWait) {
...
     });
 
-    this.Given(/^an existing pipeline$/, () => {
+    this.Given(/^an existing pipeline$/, () =>
         request({
             uri: `${this.instance}/v4/pipelines`,
             method: 'POST',
             auth: {
-                username: this.username,
+                user: this.username,
```
Comment (minz1027) [inline features/step_definitions/git-flow.js]: Don't think we need `username` here. `jwt` alone should be sufficient. Also wonder why we need it in Before hook

Comment (tkyi) [inline features/step_definitions/git-flow.js]: `username` might be used later in the authorization functional test, but I can remove it here and add it in a later pull request if it's needed.

Comment (tkyi) [inline features/step_definitions/git-flow.js]: Also removed `user`

**Comment by d2lam** [comment]
No PR job was created for this: https://api.screwdriver.cd/v4/pipelines/27cec39c9baea6cf64c684479115300d22039b2a/jobs

???

**Comment by d2lam** [comment]
Cool this is good to merge. Was trying to figure out why there wasn't a screwdriver build. @stjohnjohnson  figured out it's the github plugin issue. Nothing from us 👍 

https://github.com/screwdriver-cd/screwdriver/issues/246

**Pull request merged**